src/storage.rs

```diff
@@ -419,200 +450,324 @@ impl Cache {
-                trace!(
-                    "  downloading {}:{} to {}",
-                    name,
-                    version,
-                    download_to.display()
-                );
-                handles.push(runtime.spawn(download_package(
-                    network.clone(),
-                    name.to_string(),
-                    (*version).clone(),
-                    download_to,
-                    unpack_to,
-                )));
-            }
+                if *version == resolver::ROOT_VERSION {
+                    return Ok(root.join(CACHE_EMPTY_PACKAGE));
+                }
 
-            for handle in handles {
-                let res = runtime.block_on(handle)?;
-                res?;
-            }
-        }
+                let dir_name = format!("{}-{}", package, version);
 
-        trace!("all fetched!");
+                // First try to get a cached copy from cargo's registry.
+                if let Some(reg) = self.cargo_registry.as_ref() {
+                    let fetched_src = reg.src().join(&dir_name);
+                    if fetch_is_ok(&fetched_src).await {
+                        return Ok(fetched_src);
+                    }
+                }
 
-        Ok(paths)
+                // Paths for the fetched package and checkout in our local cache.
+                let fetched_package = root
+                    .join(CACHE_REGISTRY_CACHE)
+                    .join(format!("{}.crate", dir_name));
+                let fetched_src = root.join(CACHE_REGISTRY_SRC).join(&dir_name);
+
+                // Check if the resource is already available in our local cache.
+                let fetched_package_ = fetched_package.clone();
+                let cached_file = tokio::task::spawn_blocking(move || {
+                    File::open(&fetched_package_).map(|file| {
+                        // Update the atime and mtime for this crate to ensure it isn't
+                        // collected by the gc.
+                        let now = filetime::FileTime::now();
+                        if let Err(err) =
+                            filetime::set_file_handle_times(&file, Some(now), Some(now))
+                        {
+                            warn!(
+                                "failed to update mtime for {}, gc may not function correctly: {}",
+                                fetched_package_.display(),
+                                err
+                            );
+                        }
+                        file
+                    })
+                })
+                .await?;
+
+                // If the file isn't in our local cache, make sure to download it.
+                let file = match cached_file {
+                    Ok(file) => file,
+                    Err(_) => {
+                        let network = network
+                            .as_ref()
+                            .expect("running as --frozen but needed fetches!");
```

Comment:
Hmm now that we're doing things at this granularity we can arguably just fail this one download and any diffs that want to key off of it by storing a Result in the Once (not for this PR, followup).

Reply:
Or wait is OnceCell already doing this for us?

Reply:
because of how I'm initializing it (with the `get_or_try_init` method), if the `OnceCell` fails we'll just not initialize it, return an error, and try again next time we request it. I could also just store the result in the `OnceCell`, but that's a bit trickier, as `eyre::Report` isn't `Clone`, so this seemed like an easier approach.

Reply:
hmm ok, we'll get this in followups if/when we stop using Report so arbitarily and have our own Error types.